### DISCUSSION TITLE
VK Provider email problems

### DISCUSSION THREAD

**Discussion by xxtereshko** [nextauthjs/next-auth#1617] [Help]
**VK Provider email problem**
I can not receive email after authorization.

**What are you trying to do**
I do everything according to the documentation with example project . Email scope is configured. When authorizing on the VK side, the application requests access rights to the email. After authorization and redirection to my application, everything is fine, I even get a link to my profile avatar. But there is a problem, in the session object, the email field is null.

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

**Comment by balazsorban44** [comment]
please provide a reproduction so we can help you

**Reply by xxtereshko** [reply]
There is an example in the next-auth-example repository. I add the VK provider as in the documentation on your site.  Everything works well, but after user authentication, the session object does not contain email, although in the source code src/ provoders/vk.js has everything for this.

As I understand it, this worked at the time of the development of this provider and at some point the user's mail came. 

Configured provider's scope in accordance with vk oauth documentation. In the oauth vk authorization dialog, the user is asked for permission to send email. 

I printed the keys for the authorization object, but I did not find anything similar to email. (I just printed out all the keys of the response object. Attachment below)

![image](https://user-images.githubusercontent.com/19520678/113079201-e3732300-91d4-11eb-9b94-7d5171932b03.png)


I also opened an issue in the vk sdk repository, but I have not received a response yet.

**Comment by novavovikov** [comment]
VK uses email or phone for auth. New users in vk don't have email.

For fast solution: **You can add custom provider to your configuration**

Use bellow code:

```tsx
import {OAuthConfig, OAuthUserConfig} from 'next-auth/providers';

interface VkProfile {
    response: Array<{
        id: number;
        first_name: string;
        last_name: string;
        photo_100: string;
        can_access_closed: boolean;
        is_closed: boolean;
    }>;
}

export default function VK<P extends Record<string, any> = VkProfile>(
    options: OAuthUserConfig<P>,
): OAuthConfig<P> {
    const apiVersion = '5.126'; // https://vk.com/dev/versions

    return {
        id: 'vk',
        name: 'VK',
        type: 'oauth',
        authorization: `https://oauth.vk.com/authorize?scope=email&v=${apiVersion}`,
        token: {
            url: `https://oauth.vk.com/access_token?v=${apiVersion}`,
            async request({client, params, checks, provider}) {
                // exclude user_id and email from response
                const {user_id, email, ...tokens} = await client.oauthCallback(
                    provider.callbackUrl,
                    params,
                    checks,
                    {
                        exchangeBody: {
                            client_id: options.clientId,
                            client_secret: options.clientSecret,
                        },
                    },
                );

                return {tokens};
            },
        },
        userinfo: `https://api.vk.com/method/users.get?fields=photo_100&v=${apiVersion}`,
        profile(result: P) {
            const profile = result.response?.[0] ?? {};

            return {
                id: profile.id,
                name: [profile.first_name, profile.last_name]
                    .filter(Boolean)
                    .join(' '),
                email: null,
                image: profile.photo_100,
            };
        },
        options,
    };
}
```
